source/credentials_provider_cognito.c

```diff
@@ -0,0 +1,925 @@
+    AWS_ZERO_STRUCT(*login);
+
+    return AWS_OP_ERR;
+}
+
+static void s_aws_cognito_login_clean_up(struct aws_cognito_login *login) {
+    aws_byte_buf_clean_up(&login->login_buffer);
+}
+
+struct aws_credentials_provider_cognito_impl {
+    struct aws_http_connection_manager *connection_manager;
+    struct aws_retry_strategy *retry_strategy;
+    const struct aws_auth_http_system_vtable *function_table;
+
+    struct aws_string *endpoint;
+
+    struct aws_string *identity;
+
+    struct aws_array_list logins;
+
+    struct aws_string *custom_role_arn;
+};
+
+struct cognito_user_data {
+    struct aws_allocator *allocator;
+
+    struct aws_credentials_provider *provider;
+
+    aws_on_get_credentials_callback_fn *original_callback;
+    void *original_user_data;
+
+    struct aws_http_connection *connection;
+    struct aws_http_message *get_credentials_request;
+    struct aws_byte_buf request_body_buffer;
+    struct aws_input_stream *request_body_stream;
+
+    struct aws_retry_token *retry_token;
+    struct aws_credentials *credentials;
+    struct aws_byte_buf response_body;
+};
+
+static void s_user_data_reset(struct cognito_user_data *user_data) {
+    aws_byte_buf_clean_up(&user_data->request_body_buffer);
+    AWS_ZERO_STRUCT(user_data->request_body_buffer);
+
+    user_data->request_body_stream = aws_input_stream_release(user_data->request_body_stream);
+    user_data->get_credentials_request = aws_http_message_release(user_data->get_credentials_request);
+
+    struct aws_credentials_provider_cognito_impl *impl = user_data->provider->impl;
+    if (user_data->connection != NULL) {
+        impl->function_table->aws_http_connection_manager_release_connection(
+            impl->connection_manager, user_data->connection);
+        user_data->connection = NULL;
+    }
+
+    aws_byte_buf_reset(&user_data->response_body, false);
+}
+
+static void s_user_data_destroy(struct cognito_user_data *user_data) {
+    if (user_data == NULL) {
+        return;
+    }
+
+    s_user_data_reset(user_data);
+
+    aws_byte_buf_clean_up(&user_data->response_body);
+    aws_retry_token_release(user_data->retry_token);
+    aws_credentials_provider_release(user_data->provider);
+    aws_credentials_release(user_data->credentials);
+
+    aws_mem_release(user_data->allocator, user_data);
+}
+
+static struct cognito_user_data *s_user_data_new(
+    struct aws_credentials_provider *provider,
+    aws_on_get_credentials_callback_fn callback,
+    void *user_data) {
+
+    struct aws_allocator *allocator = provider->allocator;
+    struct cognito_user_data *cognito_user_data = aws_mem_calloc(allocator, 1, sizeof(struct cognito_user_data));
+    if (cognito_user_data == NULL) {
+        return NULL;
+    }
+
+    cognito_user_data->allocator = allocator;
+
+    if (aws_byte_buf_init(
+            &cognito_user_data->response_body, cognito_user_data->allocator, HTTP_RESPONSE_BODY_INITIAL_SIZE)) {
+        goto on_error;
```

Comment:
allocation failure is impossible now

Reply:
I just... the whole crash on allocation argument reminds me of checked exceptions.  Take this function for example.  If I remove the error checking wrt calloc/byte_buf_init then this function is now also succeed-or-crash.  So now must all the callers of this function stop checking success?  Where do you draw the line?